sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -392,4 +392,236 @@ case class GpuArrayMax(child: Expression) extends GpuBaseArrayAgg with ImplicitC
+      ColumnVector.sequence(start.getBase, ret(0), ret(1))
+    }
+  }
+
+  override def doColumnar(start: GpuScalar, stop: GpuColumnVector): ColumnVector = {
+    withResource(calculateSizeAndStep(start.getBase, stop.getBase)) { ret =>
+      withResource(ColumnVector.fromScalar(start.getBase, stop.getRowCount.toInt)) { startV =>
+        ColumnVector.sequence(startV, ret(0), ret(1))
+      }
+    }
+  }
+
+  override def doColumnar(start: GpuColumnVector, stop: GpuScalar): ColumnVector = {
+    withResource(calculateSizeAndStep(start.getBase, stop.getBase)) { ret =>
+      ColumnVector.sequence(start.getBase, ret(0), ret(1))
+    }
+  }
+
+  override def doColumnar(numRows: Int, lhs: GpuScalar, rhs: GpuScalar): ColumnVector =
+    throw new IllegalStateException("This is not supported yet")
+
+}
+
+/** GpuSequence with step */
+case class GpuSequenceWithStep(start: Expression, stop: Expression, step: Expression,
+    timeZoneId: Option[String] = None) extends GpuTernaryExpression with TimeZoneAwareExpression {
+
+  override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression =
+    copy(timeZoneId = Some(timeZoneId))
+
+  override def first: Expression = start
+
+  override def second: Expression = stop
+
+  override def third: Expression = step
+
+  override def dataType: DataType = ArrayType(start.dataType, containsNull = false)
+
+  /* TODO do we need to check below conditions?
+    (step > num.zero && start <= stop)
+        || (step < num.zero && start >= stop)
+        || (step == num.zero && start == stop),
+   */
+  private def calculateSize(
+      start: BinaryOperable,
+      stop: BinaryOperable,
+      step: BinaryOperable,
+      rows: Int): ColumnVector = {
+    // First calculate sizeWithNegative=floor((stop-start)/step)+1
+    // Second calculate size=if(sizeWithNegative < 0) 0 else sizeWithNegative
+    // Third if start==stop && step == 0, let size = 1.
+    withResource(stop.sub(start)) { difference =>
+      withResource(difference.floorDiv(step)) { quotient =>
+        withResource(Scalar.fromInt(1)) { scalarOne =>
+          withResource(quotient.add(scalarOne)) { sizeWithNegative =>
+            withResource(Scalar.fromInt(0)) { scalarZero =>
+              withResource(sizeWithNegative.greaterOrEqualTo(scalarZero)) { pred =>
+                withResource(pred.ifElse(sizeWithNegative, scalarZero)) { tmpSize =>
+                  // when start==stop, step==0, size will be 0. but we should change size to 1
+                  withResource(difference.equalTo(scalarZero)) { diffHasZero =>
+                    step match {
+                      case stepScalar: Scalar =>
+                        withResource(ColumnVector.fromScalar(stepScalar, rows)) { stepV =>
+                          withResource(stepV.equalTo(scalarZero)) { stepHasZero =>
+                            withResource(diffHasZero.and(stepHasZero)) { predWithZero =>
+                              predWithZero.ifElse(scalarOne, tmpSize)
```

Comment:
Is there any way to get rid of such nested `withResource`? Otherwise, this looks so cluster.

Reply:
Hmm, refined this.